server/src/main/scala/org/http4s/server/middleware/CSRF.scala

```diff
@@ -0,0 +1,228 @@
+    }
+
+  /** Generate a new token **/
+  private[middleware] def generateToken: F[String] =
+    signToken(CSRF.genTokenString)
+
+  /** Decode our CSRF token and extract the original token string to sign
+    */
+  private[middleware] def extractRaw(token: String): OptionT[F, String] =
+    token.split("-") match {
+      case Array(raw, nonce, signed) =>
+        OptionT(F.delay {
+          val mac = Mac.getInstance(CSRF.SigningAlgo)
+          mac.init(key)
+          val out = mac.doFinal((raw + "-" + nonce).getBytes(StandardCharsets.UTF_8))
+          if (MessageDigest.isEqual(out, Base64.getDecoder.decode(signed))) {
+            Some(raw)
+          } else {
+            None
+          }
+        })
+      case _ =>
+        OptionT.none
+    }
+
+  /** To be only used on safe methods: if the method is safe (i.e doesn't modify data),
+    * embed a new token if not present, or regenrate the current one to mitigate
+    * BREACH
+    */
+  private[middleware] def validateOrEmbed(
+      r: Request[F],
+      service: HttpService[F]): OptionT[F, Response[F]] =
+    CSRF.cookieFromHeaders(r, cookieName) match {
+      case Some(c) =>
+        OptionT.liftF(
+          (for {
+            raw <- extractRaw(c.content)
+            res <- service(r)
+            newToken <- OptionT.liftF(signToken(raw))
+          } yield res.addCookie(Cookie(name = cookieName, content = newToken)))
+            .getOrElse(Response[F](Status.Unauthorized)))
+      case None =>
+        service(r).semiflatMap(embedNew)
+    }
+
+  /** Check for CSRF validity for an unsafe action. **/
+  private[middleware] def checkCSRF(r: Request[F], service: HttpService[F]): F[Response[F]] =
+    (for {
+      c1 <- OptionT.fromOption[F](CSRF.cookieFromHeaders(r, cookieName))
+      c2 <- OptionT.fromOption[F](r.headers.get(CaseInsensitiveString(headerName)))
+      raw1 <- extractRaw(c1.content)
+      raw2 <- extractRaw(c2.value)
+      response <- if (CSRF.isEqual(raw1, raw2)) service(r) else OptionT.none
+      newToken <- OptionT.liftF(signToken(raw1)) //Generate a new token to guard against BREACH.
+    } yield response.addCookie(Cookie(name = cookieName, content = newToken)))
+      .getOrElse(Response[F](Status.Unauthorized))
+
+  /** Check method safety, then apply the correct csrf policy **/
+  private[middleware] def filter(r: Request[F], service: HttpService[F]): OptionT[F, Response[F]] =
+    if (r.method.isSafe) {
+      validateOrEmbed(r, service)
+    } else {
+      OptionT.liftF(checkCSRF(r, service))
+    }
+
+  /** Constructs a middleware that will check for the csrf token
+    * presence on both the proper cookie, and header values.
+    *
+    * If it is a valid token, it will then embed a new one,
+    * to effectively randomize the complete token while
+    * avoiding the generation of a new secure random Id, to guard
+    * against [BREACH](http://breachattack.com/)
+    *
+    */
+  def validate: HttpMiddleware[F] = { service =>
```

Comment:
So what this leaves uncovered are these cases:

```scala
case GET / "stupid" => truncate() *> Gone() // should be covered
case POST / "silly" => Ok("I could have been a GET.") // doesn't need to be covered
```

We can mitigate that if `validate` takes a `Request => Boolean`, that defaults to `_.method.isSafe`.  Or we can say, "we're not going to support the abuse of HTTP methods, and we're not going to let you shoot yourself in the foot whitelisting endpoints."

Perhaps more legitimately, some users might want CSRF checks only if an Authorization or Cookie is present.  This is the [play default](https://www.playframework.com/documentation/2.6.x/ScalaCsrf).  This could also be handled with a predicate.

I'm not strongly for it or against it, but if we think we might want it, we should get it into 0.16.

Reply:
> "we're not going to support the abuse of HTTP methods, and we're not going to let you shoot yourself in the foot whitelisting endpoints."

This is the stance I pushed with the latest commit. 

> We can mitigate that if validate takes a Request => Boolean, that defaults to _.method.isSafe

this is an interesting idea, but subject to user misuse.

> Perhaps more legitimately, some users might want CSRF checks only if an Authorization or Cookie is present. This is the play default. This could also be handled with a predicate.

We could meet half way? I think i'm not pro-Http method abuse, so we could make the condition
`if (predicate || method.isSafe) embedOrValidate else checkCSRF`.

How do you feel about this midpoint?

Reply:
Alright, on gitter we decided a `predicate: Request => Boolean = _.method.IsSafe` is a decent default.

Commenting here for the sake of documenting.